### PR TITLE
Implement min/max window dimension constraints for MacOS.

### PR DESCRIPTION
**Pull request by frewsxcv**
Fixes https://github.com/tomaka/winit/issues/80.

### CODE REVIEW

**Comment by tomaka** [comment]
As with all OSX PRs I'm not qualified to review, so hopefully someone can give their opinion.

**Comment by frewsxcv** [comment]
Example of this in action:

untitled

**Comment by mitchmindtree** [comment]
@frewsxcv nice! Could you perhaps add this example to the examples directory (then I'd be happy to test it)?

**Comment by frewsxcv** [comment]
> @frewsxcv nice! Could you perhaps add this example to the examples directory (then I'd be happy to test it)?

Done.

For the record, here's more info about Cocoa constraints:

http://stackoverflow.com/a/33696843

**Comment by mitchmindtree** [comment]
Hmmm when running the `constraints.rs` example on OS X 10.10.5 the window shows for about a frame or two before crashing with a segfault:

```
2016-11-24 01:46:21.283 constraints[19813:5427080] -[NSView widthAnchor]: unrecognized selector sent to instance 0x7f9580d24010
Segmentation fault: 11
```

The other examples seem to run fine.

**Comment by frewsxcv** [comment]
Oh, well that's not good. Looks like it was added in macOS 10.11+.

https://developer.apple.com/reference/appkit/nsview/1526638-widthanchor?language=objc

Uhhhhhhhhh

**Comment by mitchmindtree** [comment]
@frewsxcv perhaps this could be a possible alternative method?

http://stackoverflow.com/questions/5795488/setting-the-minimum-and-maximum-size-of-a-window

**Comment by frewsxcv** [comment]
I'll give that a try.

**Comment by mitchmindtree** [comment]
Also, maybe `min_max_size.rs` could be a clearer name for the example? Probably no need to print out the events in it either?

**Comment by frewsxcv** [comment]
@mitchmindtree Let me know if those changes work for you.

File: src/platform/macos/mod.rs [hunk 1]
```diff
@@ -627,6 +631,55 @@ impl Window {
...
 }
 
+unsafe fn nswindow_set_min_dimensions<V: NSWindow + Copy>(
+    window: V, min_width: f64, min_height: f64)
+{
+    window.setMinSize_(NSSize {
+        width: min_width,
+        height: min_height,
+    });
+    // If necessary, resize the window to match constraint
```
Comment (frewsxcv) [inline src/platform/macos/mod.rs]: Unlike the previous strategy, calling `setMinSize` does *not* change the window size if the window is smaller than the minimum size. In this case, if the user attempts to resize, the window will immediately expand to the minimum size, which is not a great UX, so I decided to update the window size if necessary.

**Comment by mitchmindtree** [comment]
Nice! This seems to work perfectly :)

So to clarify, the specified bounds in this case affect the *outer* size of the window (inclusive of the title bar). E.g. when setting the resize callback and making the window as small as possible in the new `min_max_size.rs` example, the window_resized callback is called with `400x178` even though the specified bounds are 400, 200. The `WindowBuilder::with_min/max_dimensions` docs don't specify whether it should bound the inner or outer points size. @tomaka is the outer_size expected in this case? Should this be documented? Maybe it is somewhere else and I missed it.

Otherwise looks good to me!

> so I decided to update the window size if necessary.

Yeah this sounds like a good call :+1:

**Comment by tomaka** [comment]
Does this work if `min_size > max_size`?

**Comment by frewsxcv** [comment]
> Does this work if min_size > max_size?

It looks like if min_size is greater, it changes the window size to the max size and doesn't allow resizing.

**Comment by tomaka** [comment]
The Android stuff could be removed, but I'll do it in another PR.

**Comment by frewsxcv** [comment]
@tomaka I don't remember if you use Windows or not; do you know if the min/max constraints work on Windows?

**Comment by tomaka** [comment]
Yes they do.

**Pull request merged**